### DISCUSSION TITLE
Define workspace outside gui-startup and open on demand

### DISCUSSION THREAD

**Discussion by aniketgm** [wezterm/wezterm#4119] [General]
Is there a way to define workspace without being opened by gui-startup ? And open as and when required (on demand), by opening the Launcher menu and opening the workspace of the thing I'm working on.

*Use-case:* I work on many issues/projects, and I have defined workspaces for all those. However, I would not be working on all those projects simultaneously. Some days I would be working on ProjA, some days ProjB, some days only issues/bugs etc. The workspaces defined inside gui-startup will open all those projects. I somewhat feel this is waste of resources and the workspaces are just there open without me switching to those.

**Comment by gegoune** [comment]
I have similar use case and solved it by attaching functions creating workspaces to some key mappings.

**Reply by aniketgm** [reply]
I'm guessing the functions are in lua. Is it possible to share ? I don't have hands on with lua at the moment.
From what I can understand from the lua config is that there needs to be some form of function provided to `action` right ?
For e.g.:
```lua
{
  key = '<some_key>',
  mods = 'ALT|CTRL|WHATEVER',
  action = <function_to_open_workspace>  # I'm not sure how to create such a function atm ?
}
```

**Reply by gegoune** [reply]
Yes, I am wrapping my function in `wezterm.action_callback`, like so

```lua
action = wezterm.action_callback(function(_win, _pane)
  local workspace_name = 'my-workspace'
  if has_value(wezterm.mux.get_workspace_names(), workspace_name) then
    wezterm.mux.set_active_workspace(workspace_name)
    return
  end

  -- open your windows, panes, etc here, note `workspace` key which causes new workspace to be created if it doesn't exist yet
  local _tab, new_pane, window = wezterm.mux.spawn_window { workspace = workspace_name, cwd = somecwd }

  wezterm.mux.set_active_workspace(workspace_name)
end)
```

This key mapping will create desired workspace or switch to it if it already exists. 

`has_value()` is a little helper function to check if table contains given value:

```lua
local has_value = function(tab, val)
  for _index, value in ipairs(tab) do
    if value == val then return true end
  end

  return false
end
```

Hope it helps, good luck! I am not sure if that's the correct way of doing it but it works. :)

**Reply by aniketgm** [reply]
Thanks a lot 👍
